ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs

```diff
@@ -0,0 +1,188 @@
+{-# LANGUAGE RecordWildCards     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TupleSections       #-}
+
+module Ouroboros.Consensus.Mempool.Impl (
+    openMempool
+  ) where
+
+import           Control.Monad.Except
+import qualified Data.Foldable as Foldable
+import           Data.Sequence (Seq (..))
+import qualified Data.Sequence as Seq
+
+import           Control.Monad.Class.MonadSTM
+
+import           Ouroboros.Network.Block (ChainHash, StandardHash)
+import qualified Ouroboros.Network.Block as Block
+
+import           Ouroboros.Storage.ChainDB.API
+
+import           Ouroboros.Consensus.Ledger.Abstract
+import           Ouroboros.Consensus.Mempool.API
+import           Ouroboros.Consensus.Util (repeatedly)
+
+{-------------------------------------------------------------------------------
+  Top-level API
+-------------------------------------------------------------------------------}
+
+openMempool :: (MonadSTM m, StandardHash blk, ApplyTx blk)
+            => ChainDB m blk hdr -> m (Mempool m blk)
+openMempool chainDB = do
+    env <- initMempoolEnv chainDB
+    return Mempool {
+        addTxs = implAddTxs env
+      , getTxs = implGetTxs env
+      }
+
+{-------------------------------------------------------------------------------
+  Internal state
+-------------------------------------------------------------------------------}
+
+-- | Internal state in the mempool
+data InternalState blk = IS {
+     -- | Transactions currently in the mempool
+      isTxs :: Seq (GenTx blk)
+
+      -- | The tip of the chain that 'isTxs' was validated against
+    , isTip :: ChainHash blk
+    }
+
+data MempoolEnv m blk hdr = MempoolEnv {
+      mpEnvChainDB  :: ChainDB m blk hdr
+    , mpEnvStateVar :: TVar m (InternalState blk)
+    }
+
+initInternalState :: InternalState blk
+initInternalState = IS Seq.empty Block.GenesisHash
+
+initMempoolEnv :: MonadSTM m => ChainDB m blk hdr -> m (MempoolEnv m blk hdr)
+initMempoolEnv chainDB = do
+    isVar <- atomically $ newTVar initInternalState
+    return $ MempoolEnv chainDB isVar
+
+{-------------------------------------------------------------------------------
+  Implementation
+-------------------------------------------------------------------------------}
+
+-- TODO: This may return some transactions as invalid that aren't new. Remove?
```

Comment:
Because we delay validating until this or `implGetTxs` are called, and in general we will have had block updates in between.

Reply:
True, but if we didn't do that, we'd _still_ be silently removing transactions (but then in the thread that monitors for chain updates).